Docs: Add documentation for the JS SDK

hugocaillard opened the pull request:
### Description

- [x] Add a feature guide about the SDK
- [x] Add an how-to guide to test contracts with the SDK
- [x] Update the rest of the docs to clearly mention the deprecation of clarinet deno features and the upcoming release of Clarinet 2

mefrem commented on docs/how-to-guides/how-to-test-contract-with-clarinet-sdk.md:
```diff
@@ -0,0 +1,516 @@
+The file `tests/counter_test.ts` that was created by `clarinet contract new counter` can be deleted.  
+
+You can have a look at `tests/contract.test.ts`, it's a sample file showing how to use the SDK with Vitest.
+It can safely be deleted.
```
I assume Clarinet will continue to make these test folders and files. Will the recommended work flow be to ask developers to delete these extraneous files?

hugocaillard replied:
Future clarinet versions will directly create the right file with the right content (`<contract-name>.test.ts`).

So this is temporary

mefrem commented on docs/how-to-guides/how-to-test-contract-with-clarinet-sdk.md:
```diff
@@ -0,0 +1,516 @@
+- The first one checks that the `increment` function returns the new value and saves it to the `count` variable.
+- The second one checks that an `print_event` is emitted when the increment function is called.
+
+> You can use `Cl.prettyPrint(value: ClarityValue)` to format any Clarity value into readable Clarity code. It can be useful to debug functions results or events values.
```
Great feature, this prettyPrint!

Also, nitpick: `Cl` is an ambiguous namespace. Is it short for Clarinet? Is it a bad idea to reserve just `C` or `clar`?

hugocaillard replied:
@mefrem `Cl` is directly imported from [Stacks.js](https://stacks.js.org/modules/_stacks_transactions.Cl)

mefrem commented on docs/how-to-guides/how-to-test-contract-with-clarinet-sdk.md:
```diff
@@ -0,0 +1,516 @@
+import { Cl } from "@stacks/transactions";
+import { describe, expect, it } from "vitest";
+
+const accounts = vm.getAccounts();
```
I think it might be worthwhile to make a link the clarinet-sdk API reference guide early on here ("Take a look at the [API reference guide] (link) for more information about the methods and elements of the clarinet-sdk").

This brings me back to one of the resources often requested by developers, which was an exhaustive list of the methods and elements available to write tests. Will VS Code provide autocomplete to provide some indication of the methods and elements available? 

The "clarinet-js-sdk.md" doc is a great start!

mefrem commented on docs/how-to-guides/how-to-test-contract-with-clarinet-sdk.md:
```diff
@@ -0,0 +1,516 @@
+
+describe("test increment method", () => {
+  it("increments the count by the given value", () => {
+    const res1 = vm.callPublicFn("counter", "increment", [Cl.uint(1)], address1);
```
Nitpick: Can we expand all the way to `result1`?

hugocaillard replied:
Sure, expanded allll the way to `incrementResponse`

mefrem commented on docs/how-to-guides/how-to-test-contract-with-clarinet-sdk.md:
```diff
@@ -0,0 +1,516 @@
+
+The response type is noted `(response <ok-type> <error-type>)` in Clarity.
+It can be `(ok <ok-type>)` or `(err <error-type>)`.
+They are called composite types, meaning that they contain an other Clarity value.
```
"another"

mefrem reviewed (changes requested):
Looks great! Made some comments, and also am curious how complete the clarinet-sdk API Reference is for developers looking to write tests with any conceivable element and method of a simnet object (blockchain, address, etc.), contract, Clarity value, etc.

ryanwaits commented:
### introduction.md

- [x] LGTM

ryanwaits commented:
### how-to-set-up-local-development-environment.md

- [ ] I think a more specific example name for the contract example, ie `counter.clar` vs `my-contract`
- [ ] The `::list_functions` does not exist as an option inside `clarinet console` in Clarinet version 1.7.1 which seems to be the latest (I see 1.8.0 in the release but I don't think I can upgrade via brew) so we maybe remove that and mention `::get_contracts` which also will display available functions
- [ ] In addition to the code example for calling a function inside `clarinet console`, above it we should provide sample code of the Clarity contract, maybe something like:
```clojure
(define-public (increment-number)
  (begin
    (var-set number 
      (+ (var-get number) u1)
    )
    (ok (var-get number))
  )
)
```
- [ ] Add code example for more explicit `contract-call` using full address:
```clojure
(contract-call? .counter increment-number) ;; (ok u1)

;; OR
(contract-call? ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.counter increment-number) ;; (ok u1)
```

ryanwaits commented:
### how-to-test-contract.md

- [x] LGTM

hugocaillard commented:
@mefrem I addressed most of your comments and replied on other.s Feel free to follow up on the answers or mark the conversation as resolved

hugocaillard commented:
Thanks a lot @ryanwaits.

I addressed most of your comments.

- I didn't perform _all_ of the changes in `how-to-set-up-local-development-environment.md` because I'd like to focus on the SDK in this PR.

- the `vm` vs `simnet` issue is most likely to be fixed. I suspect that you were testing / reviewing right when I was releasing this last minute breaking change

hugocaillard commented:
Regarding your general comment @mefrem 

>  and also am curious how complete the clarinet-sdk API Reference is for developers looking to write tests

The clarinet-sdk feature guide doesn't rely mention test,s because the SDK can be used generically to write any sort of script on top of clarinet projects, or to build tooling (we expect it to be used within clarigen for instance).

So the testing framework (which is a tool build on top of the sdk) is mostly documented in the how to guide)

Do that make sense?

👉 **Maybe we could put both in the `feature guides` so that the sit next to each other. I would probably make sense**

ryanwaits commented:
### clarinet-js-sdk.md

- [ ] Assuming that you are trying to link to `test-contract-with-clarinet-sdk.md` in `Write unit tests for Clarity smart contracts`, it's currently self referencing
- [ ] Make the code snippet `.js` compatible

ryanwaits commented on components/clarinet-sdk/README.md:
```diff
@@ -25,7 +25,9 @@ async function main() {
   const accounts = simnet.getAccounts();
-  const w1 = accounts.get("wallet_1")!;
+  const w1 = accounts.get("wallet_1");
+  if (!address1) throw new Error("invalid wallet name.");
```
typo: `address1` should be `w1`

ryanwaits reviewed (approved):
lgtm! 🚀
